### ISSUE TITLE
Bug with resolveClassOrInterfaceMembers

### ISSUE THREAD

**Issue by jbondc** [Microsoft/TypeScript#3539]
https://github.com/Microsoft/TypeScript/blob/master/src/compiler/checker.ts#L2823
```
interface monster {
	(a: "a"): void;
	(b: "b"): void;
	(c: "c"): void;
	(d: "d"): void;
	(e: "e"): void;
	(x: string): void;
}

interface m1 extends monster {
	a();
}
interface m2 extends monster {
	b();
}
interface m3 extends monster { }
interface m4 extends monster { }

interface m5 extends m1, m2, m3, m4 { }

var waa: m5;
waa("a")
```
The compiler finds 24 call expressions because it appends 'monster' from each base type.

**Comment by mhegazy** [comment]
that is by design. they should all be there, with specialized signatures all coming ahead of non specialized ones i.e. `(x: string): void;`.

**Comment by jbondc** [comment]
@mhegazy Yep but chooseOverload() on m5 ends up getting 24 'candidates' instead of 6.

**Comment by jbondc** [comment]
For better clarity:
![snap](https://cloud.githubusercontent.com/assets/199519/8229600/6764aa0c-1586-11e5-9a86-a7024ba9d3b5.png)

**Comment by jbondc** [comment]
@mhegazy Have a patch I'll send in for review later today.

**Comment by jbondc** [comment]
How is this not a bug? Still exists today.

**Comment by danquirk** [comment]
Would be nice to de-dupe the available overloads if only from a usability perspective. Would need to ensure there's no meaningful performance impact though. I'm going to call this a suggestion rather than a bug only because technically this is working correctly today and I am not 100% sure the de-duping is strictly better (namely due to the aforementioned perf concerns).

**Comment by jbondc** [comment]
I forget what case it was but performance was the reason for this patch
Somewhere:
https://github.com/Microsoft/TypeScript/blob/6db4faf4883636e70ef7a99401b9dbc5d1a58631/src/compiler/checker.ts

```chooseOverload()``` can go a bit nuts, especially with ```func("stringLiteral")``` overloading.

Much less costly to dedupe, I could take another look.

**Comment by jbondc** [comment]
Ahh may have been related to extending:
https://github.com/Microsoft/TypeScript/blob/master/lib/lib.d.ts#L6551
